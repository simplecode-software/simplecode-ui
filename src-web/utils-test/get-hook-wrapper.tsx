import React from 'react';

import { TestRenderer } from '../TestRenderer';


type TRet<T> = Readonly<{
    Wrapper: React.ComponentType<T>;
}>;

type THook<T> = (props: T) => unknown;


export const getHookWrapper = <TProps extends React.PropsWithChildren, >(
    _: THook<TProps>,
): TRet<TProps> => {
    const Wrapper: React.ComponentType<TProps> = ({
        children,
    }) => {
        return (
            <TestRenderer>
                {children}
            </TestRenderer>
        );
    };

    return {
        Wrapper,
    };
};
