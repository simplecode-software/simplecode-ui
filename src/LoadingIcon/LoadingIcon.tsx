import React from 'react';

import {
    TSimpleCodeUiContext,
    TPropsIconComponent,
} from '../types';
import { SimpleCodeUiContext } from '../SimpleCodeUiContext';


const LoadingIcon_: React.ForwardRefRenderFunction<unknown, TPropsIconComponent> = (
    props,
    ref,
) => {
    const { LoadingIcon }: TSimpleCodeUiContext = React.useContext(SimpleCodeUiContext);
    
    return (
        <LoadingIcon ref={ref} {...props} />
    );
};


export const LoadingIcon = React.memo(React.forwardRef(LoadingIcon_));
