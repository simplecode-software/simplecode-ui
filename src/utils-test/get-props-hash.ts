import { TUnknownObject } from '../types';

const asString = (value: unknown): string => {
    if (typeof value === 'string') {
        return `"${value}"`;
    }
    if (typeof value === 'number') {
        return `${value}`;
    }
    if (typeof value === 'boolean') {
        return `${value ? 'true' : 'false'}`;
    }
    if (typeof value === 'function') {
        return '[f]';
    }
    if (value instanceof Error) {
        return `"${value.message || value.constructor.name}"`;
    }
    return JSON.stringify(value);
};


const getHash = (input: string): string => {
    let hash: number = 0;
    for (let i = 0; i < input.length; i++) {
        // eslint-disable-next-line no-bitwise
        hash = input.charCodeAt(i) + ((hash << 5) - hash);
    }
    return hash.toString(16);
};

export const getPropsHash = <T extends TUnknownObject>(props: T): string => {
    const propStrings = Object.entries(props).map(([key, value]) => {
        if (value === null || value === undefined) {
            return '';
        }
        return `${key}={${asString(value)}}`;
    });
    const plainText = propStrings.filter((x) => x).join(',');
    
    return plainText ? getHash(plainText).replace(/^-/, '0') : '[EMPTY]';
};
