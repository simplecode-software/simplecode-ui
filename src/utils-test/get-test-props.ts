import { emptyArray, emptyObject } from '../constants/empty';
import { TPropVariants } from '../types';


export const getTestProps = <T extends Record<string, unknown>>(
    variants: TPropVariants<T>,
): ReadonlyArray<T> => {
    const keys: ReadonlyArray<keyof T> = Object.keys(variants);
    const values: ReadonlyArray<ReadonlyArray<T>> = Object.values(variants);
    
    const maxVariants = Math.max.apply(null, values.map((item) => item.length));
    
    let testProps: ReadonlyArray<T> = emptyArray as ReadonlyArray<T>;
    
    for (let i = 0; i < maxVariants; i++) {
        let props = emptyObject as T;
        keys.forEach((key) => {
            const values = variants[key];
            if (!values) {
                return;
            }
            const index = i % values.length;
            props = { ...props, [key]: values[index] };
        });
        
        testProps = [...testProps, props];
    }
    
    return testProps;
};
