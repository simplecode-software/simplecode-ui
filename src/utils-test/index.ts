export * from './get-props-hash';
export * from './get-test-props';
export * from './test-samples';
export * from './test-snapshots';
