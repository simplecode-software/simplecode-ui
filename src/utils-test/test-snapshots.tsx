import React from 'react';
import renderer from 'react-test-renderer';

import { TestRenderer } from '../../src-web';
import { TAnyComponent, TPropVariants } from '../types';
import { getPropsHash } from './get-props-hash';
import { getTestProps } from './get-test-props';


jest.mock('../../src-web/UiProviderWeb/ToastWeb/ToastWeb');


export const testSnapshots = <T extends Record<string, unknown>>(
    name: string,
    Component: TAnyComponent,
    variants: TPropVariants<T>,
): void => {
    describe(name, () => {
        getTestProps(variants).forEach((props) => {
            it(`renders ${getPropsHash(props)}`, () => {
                const component = renderer.create((
                    <TestRenderer>
                        <Component {...props} />
                    </TestRenderer>
                ));
                expect(component.toJSON()).toMatchSnapshot();
            });
        });
    });
};
