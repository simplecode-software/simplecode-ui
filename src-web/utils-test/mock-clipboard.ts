import { TCbVoid } from '../../src';


type TUtil = () => {
    patchClipboard: () => Clipboard;
    revertClipboard: TCbVoid;
};

export const mockClipboard: TUtil = () => {
    const originalClipboard = navigator.clipboard;

    const patchClipboard = (): Clipboard => {
        const writeText = jest.fn();
        Object.assign(navigator, {
            clipboard: {
                writeText,
            },
        });

        return navigator.clipboard;
    };
    
    const revertClipboard: TCbVoid = () => {
        Object.assign(navigator, {
            clipboard: originalClipboard,
        });
    };
    
    return {
        patchClipboard,
        revertClipboard,
    };
};
