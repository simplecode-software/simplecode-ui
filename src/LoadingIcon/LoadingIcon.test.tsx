import React from 'react';

import { testSnapshots, samplePropsLoadingIcon } from '../utils-test';
import { LoadingIcon } from './LoadingIcon';


jest.mock('../TextFragment/TextFragment');


testSnapshots('spec/LoadingIcon', LoadingIcon, samplePropsLoadingIcon);
