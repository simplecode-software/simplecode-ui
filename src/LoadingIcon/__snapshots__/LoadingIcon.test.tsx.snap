// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spec/LoadingIcon renders [EMPTY] 1`] = `
<div
  css={
    [
      [
        {
          "backgroundColor": "#c0a793",
          "borderColor": "#3918df",
        },
        null,
        null,
        null,
      ],
      {
        "alignItems": "stretch",
        "cursor": undefined,
        "display": "flex",
        "flexDirection": "column",
        "flexGrow": 0,
        "flexShrink": 0,
        "justifyContent": "flex-start",
        "opacity": 1,
        "pointerEvents": "auto",
        "position": "relative",
        "userSelect": "auto",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
    ]
  }
  disabled={null}
  href={null}
>
  <svg
    css={
      {
        "map": undefined,
        "name": "1qscsn4",
        "next": {
          "name": "animation-1xnn55h",
          "next": undefined,
          "styles": "@keyframes animation-1xnn55h{
    from {
        transform: rotate(0deg);
    }
  
    to {
        transform: rotate(720deg);
    }
}",
        },
        "styles": "
    animation: animation-1xnn55h 2s infinite;
",
        "toString": [Function],
      }
    }
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      {
        "color": undefined,
      }
    }
    viewBox="0 0 16 16"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M13.917 7A6.002 6.002 0 0 0 2.083 7H1.071a7.002 7.002 0 0 1 13.858 0h-1.012z"
      fillRule="evenodd"
    />
  </svg>
  <div
    css={
      [
        [
          {
            "backgroundColor": "#c0a793",
            "borderColor": "#3918df",
          },
          null,
          null,
          null,
        ],
        {
          "alignItems": "stretch",
          "cursor": undefined,
          "display": "flex",
          "flexDirection": "column",
          "flexGrow": 0,
          "flexShrink": 0,
          "justifyContent": "flex-start",
          "opacity": 1,
          "pointerEvents": "auto",
          "position": "relative",
          "userSelect": "auto",
        },
        [
          {
            "bottom": 0,
            "height": undefined,
            "left": 0,
            "overflow": "hidden",
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": undefined,
          },
          {
            "pointerEvents": "none",
            "position": "fixed",
            "zIndex": 1000,
          },
        ],
      ]
    }
    disabled={null}
    href={null}
  >
    <div
      css={
        [
          [
            {
              "backgroundColor": "#c0a793",
              "borderColor": "#3918df",
            },
            null,
            null,
            null,
          ],
          {
            "alignItems": "flex-end",
            "cursor": undefined,
            "display": "flex",
            "flexDirection": "column",
            "flexGrow": 0,
            "flexShrink": 0,
            "justifyContent": "flex-start",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "relative",
            "userSelect": "auto",
          },
          [
            {
              "bottom": 0,
              "left": 0,
              "overflow": "hidden",
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "bottom": 100,
              "zIndex": 100,
            },
          ],
        ]
      }
      disabled={null}
      href={null}
    />
  </div>
</div>
`;

exports[`spec/LoadingIcon renders 071f88f99 1`] = `
<div
  css={
    [
      [
        {
          "backgroundColor": "#c0a793",
          "borderColor": "#3918df",
        },
        null,
        null,
        null,
      ],
      {
        "alignItems": "stretch",
        "cursor": undefined,
        "display": "flex",
        "flexDirection": "column",
        "flexGrow": 0,
        "flexShrink": 0,
        "justifyContent": "flex-start",
        "opacity": 1,
        "pointerEvents": "auto",
        "position": "relative",
        "userSelect": "auto",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
    ]
  }
  disabled={null}
  href={null}
>
  <svg
    color="#0000ff"
    css={
      {
        "map": undefined,
        "name": "1qscsn4",
        "next": {
          "name": "animation-1xnn55h",
          "next": undefined,
          "styles": "@keyframes animation-1xnn55h{
    from {
        transform: rotate(0deg);
    }
  
    to {
        transform: rotate(720deg);
    }
}",
        },
        "styles": "
    animation: animation-1xnn55h 2s infinite;
",
        "toString": [Function],
      }
    }
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      {
        "color": "#0000ff",
      }
    }
    viewBox="0 0 16 16"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M13.917 7A6.002 6.002 0 0 0 2.083 7H1.071a7.002 7.002 0 0 1 13.858 0h-1.012z"
      fillRule="evenodd"
    />
  </svg>
  <div
    css={
      [
        [
          {
            "backgroundColor": "#c0a793",
            "borderColor": "#3918df",
          },
          null,
          null,
          null,
        ],
        {
          "alignItems": "stretch",
          "cursor": undefined,
          "display": "flex",
          "flexDirection": "column",
          "flexGrow": 0,
          "flexShrink": 0,
          "justifyContent": "flex-start",
          "opacity": 1,
          "pointerEvents": "auto",
          "position": "relative",
          "userSelect": "auto",
        },
        [
          {
            "bottom": 0,
            "height": undefined,
            "left": 0,
            "overflow": "hidden",
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": undefined,
          },
          {
            "pointerEvents": "none",
            "position": "fixed",
            "zIndex": 1000,
          },
        ],
      ]
    }
    disabled={null}
    href={null}
  >
    <div
      css={
        [
          [
            {
              "backgroundColor": "#c0a793",
              "borderColor": "#3918df",
            },
            null,
            null,
            null,
          ],
          {
            "alignItems": "flex-end",
            "cursor": undefined,
            "display": "flex",
            "flexDirection": "column",
            "flexGrow": 0,
            "flexShrink": 0,
            "justifyContent": "flex-start",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "relative",
            "userSelect": "auto",
          },
          [
            {
              "bottom": 0,
              "left": 0,
              "overflow": "hidden",
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "bottom": 100,
              "zIndex": 100,
            },
          ],
        ]
      }
      disabled={null}
      href={null}
    />
  </div>
</div>
`;

exports[`spec/LoadingIcon renders 0442ee715e 1`] = `
<div
  css={
    [
      [
        {
          "backgroundColor": "#c0a793",
          "borderColor": "#3918df",
        },
        null,
        null,
        null,
      ],
      {
        "alignItems": "stretch",
        "cursor": undefined,
        "display": "flex",
        "flexDirection": "column",
        "flexGrow": 0,
        "flexShrink": 0,
        "justifyContent": "flex-start",
        "opacity": 1,
        "pointerEvents": "auto",
        "position": "relative",
        "userSelect": "auto",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
    ]
  }
  disabled={null}
  href={null}
>
  <svg
    color="#00ff00"
    css={
      {
        "map": undefined,
        "name": "1qscsn4",
        "next": {
          "name": "animation-1xnn55h",
          "next": undefined,
          "styles": "@keyframes animation-1xnn55h{
    from {
        transform: rotate(0deg);
    }
  
    to {
        transform: rotate(720deg);
    }
}",
        },
        "styles": "
    animation: animation-1xnn55h 2s infinite;
",
        "toString": [Function],
      }
    }
    fill="currentColor"
    height={30}
    stroke="currentColor"
    strokeWidth="0"
    style={
      {
        "color": "#00ff00",
      }
    }
    viewBox="0 0 16 16"
    width={30}
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M13.917 7A6.002 6.002 0 0 0 2.083 7H1.071a7.002 7.002 0 0 1 13.858 0h-1.012z"
      fillRule="evenodd"
    />
  </svg>
  <div
    css={
      [
        [
          {
            "backgroundColor": "#c0a793",
            "borderColor": "#3918df",
          },
          null,
          null,
          null,
        ],
        {
          "alignItems": "stretch",
          "cursor": undefined,
          "display": "flex",
          "flexDirection": "column",
          "flexGrow": 0,
          "flexShrink": 0,
          "justifyContent": "flex-start",
          "opacity": 1,
          "pointerEvents": "auto",
          "position": "relative",
          "userSelect": "auto",
        },
        [
          {
            "bottom": 0,
            "height": undefined,
            "left": 0,
            "overflow": "hidden",
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": undefined,
          },
          {
            "pointerEvents": "none",
            "position": "fixed",
            "zIndex": 1000,
          },
        ],
      ]
    }
    disabled={null}
    href={null}
  >
    <div
      css={
        [
          [
            {
              "backgroundColor": "#c0a793",
              "borderColor": "#3918df",
            },
            null,
            null,
            null,
          ],
          {
            "alignItems": "flex-end",
            "cursor": undefined,
            "display": "flex",
            "flexDirection": "column",
            "flexGrow": 0,
            "flexShrink": 0,
            "justifyContent": "flex-start",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "relative",
            "userSelect": "auto",
          },
          [
            {
              "bottom": 0,
              "left": 0,
              "overflow": "hidden",
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "bottom": 100,
              "zIndex": 100,
            },
          ],
        ]
      }
      disabled={null}
      href={null}
    />
  </div>
</div>
`;

exports[`spec/LoadingIcon renders 01712ef206 1`] = `
<div
  css={
    [
      [
        {
          "backgroundColor": "#c0a793",
          "borderColor": "#3918df",
        },
        null,
        null,
        null,
      ],
      {
        "alignItems": "stretch",
        "cursor": undefined,
        "display": "flex",
        "flexDirection": "column",
        "flexGrow": 0,
        "flexShrink": 0,
        "justifyContent": "flex-start",
        "opacity": 1,
        "pointerEvents": "auto",
        "position": "relative",
        "userSelect": "auto",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
    ]
  }
  disabled={null}
  href={null}
>
  <svg
    color="#ff0000"
    css={
      {
        "map": undefined,
        "name": "1qscsn4",
        "next": {
          "name": "animation-1xnn55h",
          "next": undefined,
          "styles": "@keyframes animation-1xnn55h{
    from {
        transform: rotate(0deg);
    }
  
    to {
        transform: rotate(720deg);
    }
}",
        },
        "styles": "
    animation: animation-1xnn55h 2s infinite;
",
        "toString": [Function],
      }
    }
    fill="currentColor"
    height={20}
    stroke="currentColor"
    strokeWidth="0"
    style={
      {
        "color": "#ff0000",
      }
    }
    viewBox="0 0 16 16"
    width={20}
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M13.917 7A6.002 6.002 0 0 0 2.083 7H1.071a7.002 7.002 0 0 1 13.858 0h-1.012z"
      fillRule="evenodd"
    />
  </svg>
  <div
    css={
      [
        [
          {
            "backgroundColor": "#c0a793",
            "borderColor": "#3918df",
          },
          null,
          null,
          null,
        ],
        {
          "alignItems": "stretch",
          "cursor": undefined,
          "display": "flex",
          "flexDirection": "column",
          "flexGrow": 0,
          "flexShrink": 0,
          "justifyContent": "flex-start",
          "opacity": 1,
          "pointerEvents": "auto",
          "position": "relative",
          "userSelect": "auto",
        },
        [
          {
            "bottom": 0,
            "height": undefined,
            "left": 0,
            "overflow": "hidden",
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": undefined,
          },
          {
            "pointerEvents": "none",
            "position": "fixed",
            "zIndex": 1000,
          },
        ],
      ]
    }
    disabled={null}
    href={null}
  >
    <div
      css={
        [
          [
            {
              "backgroundColor": "#c0a793",
              "borderColor": "#3918df",
            },
            null,
            null,
            null,
          ],
          {
            "alignItems": "flex-end",
            "cursor": undefined,
            "display": "flex",
            "flexDirection": "column",
            "flexGrow": 0,
            "flexShrink": 0,
            "justifyContent": "flex-start",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "relative",
            "userSelect": "auto",
          },
          [
            {
              "bottom": 0,
              "left": 0,
              "overflow": "hidden",
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "bottom": 100,
              "zIndex": 100,
            },
          ],
        ]
      }
      disabled={null}
      href={null}
    />
  </div>
</div>
`;

exports[`spec/LoadingIcon renders 7df19aa5 1`] = `
<div
  css={
    [
      [
        {
          "backgroundColor": "#c0a793",
          "borderColor": "#3918df",
        },
        null,
        null,
        null,
      ],
      {
        "alignItems": "stretch",
        "cursor": undefined,
        "display": "flex",
        "flexDirection": "column",
        "flexGrow": 0,
        "flexShrink": 0,
        "justifyContent": "flex-start",
        "opacity": 1,
        "pointerEvents": "auto",
        "position": "relative",
        "userSelect": "auto",
      },
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
    ]
  }
  disabled={null}
  href={null}
>
  <svg
    css={
      {
        "map": undefined,
        "name": "1qscsn4",
        "next": {
          "name": "animation-1xnn55h",
          "next": undefined,
          "styles": "@keyframes animation-1xnn55h{
    from {
        transform: rotate(0deg);
    }
  
    to {
        transform: rotate(720deg);
    }
}",
        },
        "styles": "
    animation: animation-1xnn55h 2s infinite;
",
        "toString": [Function],
      }
    }
    fill="currentColor"
    height={20}
    stroke="currentColor"
    strokeWidth="0"
    style={
      {
        "color": undefined,
      }
    }
    viewBox="0 0 16 16"
    width={20}
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M13.917 7A6.002 6.002 0 0 0 2.083 7H1.071a7.002 7.002 0 0 1 13.858 0h-1.012z"
      fillRule="evenodd"
    />
  </svg>
  <div
    css={
      [
        [
          {
            "backgroundColor": "#c0a793",
            "borderColor": "#3918df",
          },
          null,
          null,
          null,
        ],
        {
          "alignItems": "stretch",
          "cursor": undefined,
          "display": "flex",
          "flexDirection": "column",
          "flexGrow": 0,
          "flexShrink": 0,
          "justifyContent": "flex-start",
          "opacity": 1,
          "pointerEvents": "auto",
          "position": "relative",
          "userSelect": "auto",
        },
        [
          {
            "bottom": 0,
            "height": undefined,
            "left": 0,
            "overflow": "hidden",
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": undefined,
          },
          {
            "pointerEvents": "none",
            "position": "fixed",
            "zIndex": 1000,
          },
        ],
      ]
    }
    disabled={null}
    href={null}
  >
    <div
      css={
        [
          [
            {
              "backgroundColor": "#c0a793",
              "borderColor": "#3918df",
            },
            null,
            null,
            null,
          ],
          {
            "alignItems": "flex-end",
            "cursor": undefined,
            "display": "flex",
            "flexDirection": "column",
            "flexGrow": 0,
            "flexShrink": 0,
            "justifyContent": "flex-start",
            "opacity": 1,
            "pointerEvents": "none",
            "position": "relative",
            "userSelect": "auto",
          },
          [
            {
              "bottom": 0,
              "left": 0,
              "overflow": "hidden",
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "bottom": 100,
              "zIndex": 100,
            },
          ],
        ]
      }
      disabled={null}
      href={null}
    />
  </div>
</div>
`;
